### ISSUE TITLE
If NIF calls system command, it keeps blocking or crashes

### ISSUE THREAD

**Issue by jackalcooper** [rusterlium/rustler#446]
To reproduce:
- clone https://github.com/jackalcooper/rustler_subprocess
```bash
git clone https://github.com/jackalcooper/rustler_subprocess.git
cd rustler_subprocess
```
- run:
```bash
mix deps.get
mix test
```

**Comment by jackalcooper** [comment]
```elixir
.thread '<unnamed>' panicked at 'failed to execute process: Os { code: 10, kind: Uncategorized, message: "No child processes" }', src/lib.rs:9:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread '<unnamed>' panicked at 'failed to execute process: Os { code: 10, kind: Uncategorized, message: "No child processes" }', src/lib.rs:19:10


  1) test dirty (RustlerSubprocessTest)
     test/rustler_subprocess_test.exs:13
     ** (ErlangError) Erlang error: :nif_panicked
     code: assert RustlerSubprocess.NIF.add_dirty(1, 2) == 3
     stacktrace:
       (rustler_subprocess 0.1.0) RustlerSubprocess.NIF.add_dirty(1, 2)
       test/rustler_subprocess_test.exs:14: (test)



  2) test normal (RustlerSubprocessTest)
     test/rustler_subprocess_test.exs:9
     ** (ErlangError) Erlang error: :nif_panicked
     code: assert RustlerSubprocess.NIF.add(1, 2) == 3
     stacktrace:
       (rustler_subprocess 0.1.0) RustlerSubprocess.NIF.add(1, 2)
       test/rustler_subprocess_test.exs:10: (test)
```

**Comment by evnu** [comment]
This looks tricky, and this might be more of an ERTS problem.

**TL;DR**:  `wait4()` signals `ECHILD` when the parent waits for the terminated child, but we would expect it to succeed.

Looking into `strace` with your [example](https://github.com/jackalcooper/rustler_subprocess), we can determine that the spawned child process succeeds.

```
$ strace -f mix test test/rustler_subprocess_test.exs:13 &> strace.log
```

In my trace, `13134` is the new child process (resulting pid from `clone` and it calls into `execve`), and `13025` is the parent:

```
strace: Process 13134 attached
...
[pid 13134] execve("/usr/bin/sh", ["sh", "-c", "echo hello"], 0x7ffe544a7e38 /* 43 vars */ <unfinished ...>
[pid 13025] <... clone resumed>)        = 13134
[pid 13025] munmap(0x7f66686d3000, 36864) = 0
[pid 13025] rt_sigprocmask(SIG_SETMASK, [],  <unfinished ...>
[pid 13134] <... execve resumed>)       = 0
```

So: `execve` succeeded for `13134`, and `clone` succeeded for `13025`. After that, `13134` loads shared libs such as `libc`, does all the things a starting process has to do on Linux,  and finally writes `hello` to `stdout`:

```
[pid 13134] write(1, "hello\n", 6)      = 6
[pid 13025] <... poll resumed>)         = 1 ([{fd=18, revents=POLLIN}])
[pid 13134] rt_sigprocmask(SIG_BLOCK, [CHLD],  <unfinished ...>
[pid 13025] read(18,  <unfinished ...>
[pid 13134] <... rt_sigprocmask resumed>[], 8) = 0
[pid 13025] <... read resumed>"hello\n", 32) = 6
[pid 13134] rt_sigprocmask(SIG_SETMASK, [],  <unfinished ...>
[pid 13025] read(18,  <unfinished ...>
[pid 13134] <... rt_sigprocmask resumed>NULL, 8) = 0
[pid 13025] <... read resumed>0x7f65d8002e46, 26) = -1 EAGAIN (Resource temporarily unavailable)
[pid 13134] exit_group(0 <unfinished ...>
[pid 13025] poll([{fd=18, events=POLLIN}, {fd=20, events=POLLIN}], 2, -1 <unfinished ...>
[pid 13134] <... exit_group resumed>)   = ?
[pid 13025] <... poll resumed>)         = 2 ([{fd=18, revents=POLLHUP}, {fd=20, revents=POLLHUP}])
[pid 13134] +++ exited with 0 +++
```

`13134` finishes successfully here. But then the parent aborts `13025`, as `wait4` returns an error:

```
[pid 13025] read(18, "", 58)            = 0
[pid 13025] ioctl(20, FIONBIO, [0])     = 0
[pid 13025] read(20, "", 32)            = 0
[pid 13025] close(20)                   = 0
[pid 13025] close(18)                   = 0
[pid 13025] wait4(13134, 0x7f66243ff504, 0, NULL) = -1 ECHILD (No child processes)
[pid 13025] write(2, "thread '", 8thread ')     = 8
[pid 13025] write(2, "<unnamed>", 9<unnamed>)    = 9
[pid 13025] write(2, "' panicked at '", 15' panicked at ') = 15
```

`wait4` returns `ECHILD`, but should succeed. In a normal rust program, `wait4` also happens and is succesfull:

```rust
fn main() {
    std::process::Command::new("sh")
        .arg("-c")
        .arg("echo hello")
        .output()
        .expect("failed to execute process");
}
```

strace:

```
wait4(27190, [{WIFEXITED(s) && WEXITSTATUS(s) == 0}], 0, NULL) = 27190
sigaltstack({ss_sp=NULL, ss_flags=SS_DISABLE, ss_size=8192}, NULL) = 0
munmap(0x7facc93bc000, 12288)           = 0
exit_group(0)                           = ?
+++ exited with 0 +++
```

I wonder why `strace` is capable of describing the `wstatus` input properly for this example, maybe that hints at the reason.

**Comment by evnu** [comment]
Side-note: If possible, you should probably spawn processes from Elixir itself, and not from a NIF. See `System.cmd/2,3` for example.

**Comment by jackalcooper** [comment]
I found that trapping the signal seems to work, diff here: https://github.com/jackalcooper/rustler_subprocess/pull/1/files

**Comment by evnu** [comment]
Interesting. See http://erlang.org/pipermail/erlang-questions/2020-November/100109.html for more information on erlang and SIGCHLD:

> After some debugging, we found out the root cause is because the Erlang VM
sets SIGCHLD to SIG_IGN.

Also from that thread:

> If the calling process sets SIGCHLD to SIG_IGN, and the process has no
unwaited for children that were transformed into zombie processes, the
calling thread blocks until all of the children of the process terminate,
at which time waitpid() returns -1 with errno set to ECHILD.

I will close this issue, setting the signal handler should be the right approach to handle this. The phenomenon is not caused by rustler, so we cannot do more here.

**Comment by listx** [comment]
@jackalcooper Can you post your code samples as a gist for posterity?

**Comment by jackalcooper** [comment]
> @jackalcooper Can you post your code samples as a gist for posterity?

oops it was deleted. Have you resolved the issue? 

That's what is left and should be similar to the original example.

```
    parent = self()

    {:ok, :dirty} =
      System.trap_signal(:sigchld, :dirty, fn ->
        send(parent, {:nif, :ok})
        :ok
      end)

    # This create sub process so we trap the signal
    declarations = Exotic.CodeGen.parse_header(h)

    receive do
      {:nif, :ok} ->
        :ok = System.untrap_signal(:sigchld, :dirty)
    end

    declarations
```

**Comment by listx** [comment]
> Have you resolved the issue?

No, I basically ran into the same problem and ended up just calling the system command from Elixir, not the NIF.

> That's what is left and should be similar to the original example.

I actually thought you did the signal trapping from inside the Rust NIF code, but I guess not.
